### PR TITLE
Northbound Cluster API

### PR DESCRIPTION
**Pull request by sbernard31**
(work in progess)

### CODE REVIEW

File: leshan-server-cluster/src/main/java/org/eclipse/leshan/server/demo/cluster/serialization/ClientUpdateSerDes.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import org.eclipse.leshan.LinkObject;
+import org.eclipse.leshan.server.client.ClientUpdate;
+
+import com.eclipsesource.json.Json;
+import com.eclipsesource.json.JsonArray;
+import com.eclipsesource.json.JsonObject;
+
+/**
+ * Functions for serialize and deserialize a ClientUpdate in JSON for storage.
```
Comment (jvermillard) [inline leshan-server-cluster/src/main/java/org/eclipse/leshan/server/demo/cluster/serialization/ClientUpdateSerDes.java]: storage?

**Comment by jvermillard** [comment]
some comments:

1. I would like to have a topic prefix, like "leshan." so we can namespace all the leshan PUB/SUB traffic, because the broker is probably shared for multiple systems.
2. Having some unique abstraction called "EventDispatcher" or something alike would make it easier to make pluggable implementation of the dispatcher using different technology

**Comment by balsmn** [comment]
I had took a look in to this PR. If I understand correctly, the JSON request message structure passed via Redis looks like 

```
{
    "ticket": xxxx
    "ep": xxxxx
    "req": {
            ... json representation of LWM2M DownlinkRequest ....
   }
}
```
I was wondering whether you also considered multi-tenancy for the northbound API or may be you have it for the future PRs.

**Comment by sbernard31** [comment]
The documentation is available [here](https://github.com/eclipse/leshan/wiki/Cluster#redis).
There is a discussion about it on the [mailing list](https://dev.eclipse.org/mhonarc/lists/leshan-dev/msg00562.html)

About multi-tenancy did you see any specific point in the implementation which could be a problem for multi-tenancy architecture ?

**Comment by balsmn** [comment]
Yes, I think so. At first glance, the tenant meta-data, say tenant id is not part of the request/response. But thats something that can be added later too. Not a big issue.

Part of Multi-tenancy is also to ensure that the response for a request doesn't reach a wrong tenant. As Redis pub/sub has no relation to key space, we have only limited options. For example, this could be achieved having separate channel prefix for each tenant. As far as I can see, the channels defined (for REQUEST_CHANNEL and RESPONSE_CHANNEL) in a generic way without any tenant relation. right?

**Comment by sbernard31** [comment]
You're right. I think multi-tenancy is cleary not in the scope for now.

**Comment by ming2377** [comment]
Really hope this pull-request will be processed soon... Thanks in advance.

**Comment by sbernard31** [comment]
This PR is awaiting for a review, but keep in mind this is still an experimental feature.

**Comment by sbernard31** [comment]
I rebased it on master.

**Comment by ming2377** [comment]
Thanks for the rebase.

File: leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisTokenHandler.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import org.eclipse.leshan.server.client.Client;
+import org.eclipse.leshan.server.client.ClientRegistryListener;
+import org.eclipse.leshan.server.client.ClientUpdate;
+import org.eclipse.leshan.util.Validate;
+
+import redis.clients.jedis.Jedis;
+import redis.clients.util.Pool;
+
+/**
+ * Only one cluster instance can be responsible of a given LWM2M client. (This restriction is mainly due to the DTLS
```
Comment (jvermillard) [inline leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisTokenHandler.java]: nly one cluster instance can be responsible of a given LWM2M client. ""at a given moment"" ?

File: leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisTokenHandler.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * session)</br>
+ * This class store the couple Cluster instance / LwM2M client in a Redis Store.</br>
+ * Each Cluster instance is identify by a unique UI and each device by its endpoint.
```
Comment (jvermillard) [inline leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisTokenHandler.java]: identified

File: leshan-server-cluster/src/main/resources/log4j2.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jvermillard) [inline leshan-server-cluster/src/main/resources/log4j2.xml]: should we let the user bring his own log config y?

**Comment by sbernard31** [comment]
This is now integrated in master but this is still very experimental and maybe will not be deliver in the first release of Leshan.

**Comment by jvermillard** [comment]
R.I.P. my review :sob:

**Pull request closed without merging**